src/Bicep.Core/Syntax/SyntaxTreeFactory.cs

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using Azure.Deployments.Core.Definitions.Schema;
+using Azure.Deployments.Core.Json;
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Extensions;
+using Bicep.Core.FileSystem;
+using Bicep.Core.Parsing;
+using Bicep.Core.Syntax.Converters;
+using Bicep.Core.Text;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Bicep.Core.Syntax
+{
+    public static class SyntaxTreeFactory
+    {
+        public static SyntaxTree CreateSyntaxTree(Uri fileUri, string fileContents, bool isEntryFile) => isEntryFile
+            ? CreateBicepSyntaxTree(fileUri, fileContents)
+            : CreateSyntaxTree(fileUri, fileContents);
+
+        public static SyntaxTree CreateSyntaxTree(Uri fileUri, string fileContents) =>
+            PathHelper.HasBicepExtension(fileUri) || !PathHelper.HasAnyExtension(fileUri)
+                ? CreateBicepSyntaxTree(fileUri, fileContents)
+                : CreateJsonSyntaxTree(fileUri, fileContents);
+
+        public static SyntaxTree CreateBicepSyntaxTree(Uri fileUri, string fileContents)
+        {
+            var parser = new Parser(fileContents);
+            var lineStarts = TextCoordinateConverter.GetLineStarts(fileContents);
+            
+            return new SyntaxTree(fileUri, lineStarts, parser.Program());
+        }
+
+        public static JsonSyntaxTree CreateJsonSyntaxTree(Uri fileUri, string fileContents)
+        {
+            var lineStarts = TextCoordinateConverter.GetLineStarts(fileContents);
+
+            try
+            {
+                var templateObject = JObject.Parse(fileContents, new JsonLoadSettings
+                {
+                    CommentHandling = CommentHandling.Ignore,
+                    LineInfoHandling = LineInfoHandling.Load,
+                });
+
+                var programSyntax = ProgramSyntaxConverter.ConvertToProgramSyntaxWithTypeInfoOnly(templateObject);
+
+                if (!Template.TryFromJson<Template>(fileContents, out var _))
+                {
+                    return CreateErrorJsonSyntaxTree(fileUri, lineStarts, "Failed to desirialize the template.", templateObject);
+                }
+
+                return new JsonSyntaxTree(fileUri, lineStarts, templateObject, programSyntax);
+            }
+            catch (JsonReaderException e)
+            {
+                return CreateErrorJsonSyntaxTree(fileUri, lineStarts, e.Message, e.LineNumber, e.LinePosition);
+            }
+            catch (SyntaxConversionException e)
+            {
+                return CreateErrorJsonSyntaxTree(fileUri, lineStarts, e.Message, e.LineNumber, e.LinePosition);
+            }
+        }
+
+        private static JsonSyntaxTree CreateErrorJsonSyntaxTree(Uri fileUri, ImmutableArray<int> lineStarts, string errorMessage, int errorLineNumber, int errorLinePosition)
```

Comment:
>CreateErrorJsonSyntaxTree [](http://example.com/codeflow?start=38&length=25)

When does the user see the error that shows up here?

Reply:
And do we ever get into a situation where the built-in JSON language server in VS code or ARM tools are giving duplicate or similar errors?

Reply:
The error will show up upon opening a JSON template that is referenced in a bicep file as a module, and that bicep file has been opened by the user before. Any other JSON files will be ignored by the language server (see changes in `BicepCompilationManager`). 

For the 2nd question - yes, users may get duplicate/similar errors for parameters and outputs if both the bicep and ARM tools extensions are installed. Maybe we should expose a VS Code config to let user suppress JSON template errors emitted by us. I'm open to submit a separate PR to implement that.

Reply:
Most of our users will have both extensions installed for a long time, so we'd have to disable this in the default experience. (Both extensions are also part of the Azure extension pack.) In general, I'm not sure we need to provide an amazing validation experience for JSON templates at all given that there is an extension that covers that in more depth than we would here.

Reply:
I agree that we don't need to worry about VS Code users, but I'm wondering if we should still provide errors for JSON templates when running `bicep build`. If we are okay with just saying "The referenced JSON template module has errors" then I'm totally fine with not emitting errors for JSON templates.

Reply:
As we discussed offline, I've changed the implementation to not show errors for JSON templates.